src/main/java/de/medizininformatikinitiative/torch/ConsentHandler.java

```diff
@@ -195,6 +199,94 @@ public Flux<Map<String, Map<String, List<ConsentPeriod>>>> buildingConsentInfo(S
+                .doOnNext(encounter -> logger.debug("Encounter fetched: {}", encounter.getIdElement().getIdPart()))
+                .onErrorResume(e -> {
+                    logger.error("Error fetching encounters for batch: {}", batch, e);
+                    return Flux.empty();
+                });
+
+        // Step 2: Group the encounters by patient ID
+        Mono<Map<String, Collection<Encounter>>> encountersByPatientMono = allEncountersFlux
+                .flatMap(encounter -> {
+                    try {
+                        String patientId = ResourceUtils.getPatientId(encounter);
+                        // Use Map.Entry to pair patient ID with the encounter
+                        return Mono.just(new AbstractMap.SimpleEntry<>(patientId, encounter));
+                    } catch (PatientIdNotFoundException e) {
+                        logger.error("Patient ID not found in encounter resource", e);
+                        return Mono.empty(); // Skip this encounter
+                    }
+                })
+                .collectMultimap(
+                        Map.Entry::getKey, // Key Mapper: patient ID
+                        Map.Entry::getValue // Value Mapper: Encounter object
+                );
+
+        // Step 3: Process each patient's consent info individually
+        return encountersByPatientMono.flatMapMany(encountersByPatientMap ->
+                consentInfoFlux.flatMap(patientConsentInfo -> {
+                    // Extract the patient ID from the consent info map
+                    String patientId = patientConsentInfo.keySet().stream().findFirst().orElse(null);
+
+                    if (patientId == null) {
+                        logger.warn("Patient ID not found in consent info");
+                        return Mono.just(patientConsentInfo); // Or handle as appropriate
+                    }
+
+                    // Get the encounters for this patient
+                    List<Encounter> patientEncounters = (List<Encounter>) encountersByPatientMap.get(patientId);
+
+                    if (patientEncounters == null || patientEncounters.isEmpty()) {
+                        logger.info("No encounters found for patient {}", patientId);
+                        // No encounters for this patient, return the consent info as is
+                        return Mono.just(patientConsentInfo);
+                    }
+
+                    // Process the encounters and update the consent periods
+                    return Mono.fromCallable(() -> {
+                        // Update the consent periods for this patient
+                        updateConsentPeriodsForPatient(patientConsentInfo, patientEncounters);
+                        return patientConsentInfo;
+                    }).subscribeOn(Schedulers.boundedElastic());
+                })
+        );
+    }
+
+    // Helper method to update consent periods for a patient based on their encounters
```

Comment:
@Lucas0T  remove comment - rename function to
updateConsentPeriodsByPatientEncounters